### PR TITLE
Major refactor, redesign of the constraint language, and working DialectType

### PR DESCRIPTION
**Pull request by nhaehnle**
This all started out with wanting to support a real `DialectType` on top of the recently added `llvm::TargetExtType` (which hasn't yet made it to all our internal LLVM branches, but that's only a matter of time). From that goal, things escalated a bit.

Basically, after a bit of refactoring I hit a snag in that we'd want to express constraints on operation results and arguments that themselves refer to type arguments. This is necessary to express an operation that behaves analogous to `extractelement` or `trunc`. What we had previously was hard to extend and generally felt a bit unprincipled.

I ended up with the last commit of this series as a giant change which essentially rewrites the constraint language.

The design superficially looks quite similar to what we had, but it feels much cleaner to me under the hood. There is a clear notion of *predicates* and of *constraints* built up out of applications of those predicates. There's more in the final commit message and in the newly added `docs/Constraints.md`. As a nice bonus, `BuiltinType` is now completely implemented in TableGen.

Oh, this change also adds verifier methods. Integrating that into a proper module verifier pass is left for the future, though.

### CODE REVIEW

**Comment by nhaehnle** [comment]
Ping. This is very large, but at least having a look at the doc and .td files would be good.

**Review by Flakebi — APPROVED** [review]
LGTM, only skimmed over the implementation in the last commit.

**Review by jasilvanus — APPROVED** [review]
I skimmed parts of the implementation, and read the tablegen doc.
Minor comments inline.

Having the possiblity to create custom types that are not touched by optimization passes seems really interesting.

File: lib/TableGen/GenDialect.cpp [hunk 1]
```diff
@@ -250,8 +256,204 @@ class Builder;
...
+  FmtContextScope scope{fmt};
+  fmt.withContext("context");
+
+  out << tgfmt(R"(
+    bool $_op::verifier(::llvm::raw_ostream &errs) {
+      ::llvm::LLVMContext &$_context = getModule()->getContext();
+      (void)$_context;
+
+      if (arg_size() != $0) {
+        errs << "  wrong number of arguments: " << arg_size()
```
[the remark this answers is not in the document]
Comment (nhaehnle) [reply lib/TableGen/GenDialect.cpp]: The indentation is intended to help make sense of the output of the planned verifier pass, where you may see verification errors for multiple instructions.

The overall output (which I'm sorry is not tested yet, I'm going to get to that later) looks like:
```
Verifier error in: <instruction here>
  wrong number of arguments: <n>, expected <m>
```

File: include/llvm-dialects/TableGen/Operations.h [hunk 2]
```diff
@@ -84,6 +84,8 @@ class Operation {
   std::vector<OpNamedValue> results;
   std::vector<std::unique_ptr<PredicateExpr>> verifier;
 
+  ~Operation();
```
[the remark this answers is not in the document]
Comment (jasilvanus) [reply include/llvm-dialects/TableGen/Operations.h:86]: To clarify: Actually I'd prefer a comment on the *declaration*, so when reading the class it is clear the destructor does not contain any nontrivial logic breaking e.g. copy semantics. But it is not that important, and I understand the potential of such a comment becoming outdated.

**Pull request merged**